DIGITfiles/SequencesFromBlast.py

```diff
@@ -181,32 +191,53 @@ def writeToBestQueriesFile(listQueries, flankseq):
+    return
 
 
 def getBestQuery(genome, query):
     '''
-    Sort hits to identify best hit for a query. Find the percent difference
-    between the best and second best hits.
+    Sort hits to identify best hit for a query. Find the percent difference between the best and
+    second best hits.
     '''
-    bestQuery = allQueries[genome][query][0]
-    secondBestQuery = None
-    for i in range(1, len(allQueries[genome][query])):
-        if allQueries[genome][query][i].bitScore > bestQuery.bitScore:
-            secondBestQuery = bestQuery
-            bestQuery = allQueries[genome][query][i]
-        elif allQueries[genome][query][i].bitScore < bestQuery.bitScore:
+    bestQuery = allQueries[genome][query][0]  # Query
+    secondBestQuery = None  # Query
+    for i in range(1, len(allQueries[genome][
+                              query])):  # iterate through Query objects in allQueries[genome][
+        # allele]
+        ### NEW: changed > to >= make sure it works before deleting this comment!!!! 5/22 ###
+        ### NOTE: this does change the output from before. yikes. still valid but the results are
+        # inconsistant...going to put the = on the "less than" comparison
+        ### note again: that fixed it.
+        if allQueries[genome][query][
+            i].bitScore > bestQuery.bitScore:  # if the new bit score is better than bestQuery's
+            # bit score
+            secondBestQuery = bestQuery  # redefine secondBestQuery as Query in bestQuery
+            bestQuery = allQueries[genome][query][i]  # redefine bestQuery as the current Query
+
+        elif allQueries[genome][query][
+            i].bitScore <= bestQuery.bitScore:  # if the new bit score is with than bestQuery's
+            # bit score
+            ###  NEW: changed > to >= make sure it works before deleting this comment!!!! 5/22 ###
             if not secondBestQuery or allQueries[genome][query][
-                i].bitScore > secondBestQuery.bitScore:
-                secondBestQuery = allQueries[genome][query][i]
+                i].bitScore >= secondBestQuery.bitScore:  # if either (there is no
+                # secondBestQuery) OR (new query's bit score is better than the secondBestWuery
+                # bit score)
+                secondBestQuery = allQueries[genome][query][
+                    i]  # redefine second best query as the new query
     # TODO: come back and make this work
-    # if second_best_query:
-    #    best_query.diff = abs(
-    #        ((best_query.bit_score - second_best_query.bit_score) / (best_query.bit_score + second_best_query.bit_score)) / 2) * 100
+    perDiff = -1
+    if secondBestQuery:  # if there is a second-best query
+        perDiff = abs(
+            (
+                    ((bestQuery.bitScore - secondBestQuery.bitScore) / (
+                                bestQuery.bitScore + secondBestQuery.bitScore)) / 2
+            ) * 100
+        )
 
     for i in range(0, len(allQueries[genome][query])):
```

Comment:
If allQueries[][] list is very large, consider caching the index so you don't have to search the list twice. Then y ou can go right to the best and update it directly. Of course, you would have to use the for loop in order to know the index.

Reply:
what do you mean by caching the index? the lists tend to be very short (1 to 5 Query objects) with a couple that have like 20 or 50